contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -0,0 +1,861 @@
+
+```golang
+// ConversionReview describes a conversion request/response.
+type ConversionReview struct {
+  metav1.TypeMeta
+  // Request describes the attributes for the conversion request.
+  // +optional
+  Request *ConversionRequest
+  // Response describes the attributes for the conversion response.
+  // +optional
+  Response *ConversionResponse
+}
+
+type ConversionRequest struct {
+  // UID is an identifier for the individual request/response. Useful for logging.
+  UID types.UID
+  // The version to convert given object to. E.g. "stable.example.com/v1"
+  APIVersion string
+  // Object is the CRD object to be converted.
+  Object runtime.RawExtension
+}
+
+type ConversionResponse struct {
+  // UID is an identifier for the individual request/response.
+  // This should be copied over from the corresponding ConversionRequest.
+  UID types.UID
+  // ConvertedObject is the converted version of request.Object.
+  ConvertedObject runtime.RawExtension
+}
+```
+
+If the conversion is failed, the webhook should fail the HTTP request with a proper error code and message that will be used to create a status error for the original API caller.
+
+
+## Monitorability
+
+There should be prometheus variables to show:
+
+* CRD conversion latency
+    * Overall
+    * By webhook name
+    * By request (sum of all conversions in a request)
+    * By CRD
+* Conversion Failures count
+    * Overall
+    * By webhook name
+    * By CRD
+* Timeout failures count
+    * Overall
+    * By webhook name
+    * By CRD
+
+Adding a webhook dynamically adds a key to a map-valued prometheus metric. Webhook host process authors should consider how to make their webhook host monitorable: while eventually we hope to offer a set of best practices around this, for the initial release we won’t have requirements here.
+
+
+## Error Messages
+
+When a conversion webhook fails, e.g. for the GET operation, then the error message from the apiserver to its client should reflect that conversion failed and include additional information to help debug the problem. The error message and HTTP error code returned by the webhook should be included in the error message API server returns to the user.  For example:
+
+```bash
+$ kubectl get mykind somename
+error on server: conversion from stored version v1 to requested version v2 for somename: "408 request timeout" while calling service "mywebhookhost.somens.cluster.local:443"
+```
+
+
+For operations that need more than one conversion (e.g. LIST), no partial result will be returned. Instead the whole operation will fail the same way with detailed error messages. To help debugging these kind of operations, the UID of the first failing conversion will also be included in the error message. 
```

Comment:
original comments need to be addressed:

by @enisoc:
```
Is this what currently happens if one of the compiled-in conversion functions returns an error?

Also, this seems to imply that conversion hooks must not have stricter expectations than validation. Otherwise, one "valid" but unconvertible object could cause denial of service to many other objects.

Do we ever check that an object in a CR is valid, other than during the admission chain? That is, do we validate objects that we read from storage?
```
by @erictune:
```
What does List do if one object fails conversion? If it return a partial list? If so, I argue that is unexpected and probably going to cause bugs. If not, then we have the same DoS problem.
```

Reply:
Please clarify whether the webhook is supposed to accept an entire List or if it will always be called for each individual item-- I expect the latter, but it needs to be explicitly stated.